svgnative/src/ports/gdiplus/GDIPlusSVGRenderer.cpp

```diff
@@ -223,14 +223,14 @@ void GDIPlusSVGRenderer::Save(const GraphicStyle& graphicStyle)
 
     if (graphicStyle.clippingPath)
     {
-        // TODO
-        SVG_ASSERT_MSG(false, "Need to implement svg clipping path!");
-    }
-
-    if (graphicStyle.opacity < 1.0f)
-    {
-        // TODO
-        SVG_ASSERT_MSG(false, "Need to implement svg opacity!");
+        std::unique_ptr<Gdiplus::GraphicsPath> clip_path(dynamic_cast<const GDIPlusSVGPath*>(graphicStyle.clippingPath->path.get())->GetGraphicsPath().Clone());
+        if (graphicStyle.clippingPath->transform)
+        {
+            const Gdiplus::Matrix* matrix = &dynamic_cast<GDIPlusSVGTransform*>(graphicStyle.clippingPath->transform.get())->GetMatrix();
+            clip_path->Transform(matrix);
+        }
+        clip_path->SetFillMode(graphicStyle.clippingPath->clipRule == WindingRule::kNonZero ? Gdiplus::FillMode::FillModeWinding : Gdiplus::FillMode::FillModeAlternate);
+        mContext->SetClip(clip_path.get(), Gdiplus::CombineModeIntersect);
```

Comment:
@dirkschulze Thanks for pointing out about clipping working differently with Gdiplus. That's true. In fact, it allows you to specify how you want to form the new clipping region. Other rendering engines such as Cairo form the new clipping region by intersection the passed path with the old one and by passing this flag, GDI+ should do the same. The SVG document you provided gives the following rendering:
![image](https://user-images.githubusercontent.com/23481137/116740321-4e9b5980-aa0e-11eb-8d5e-c0868c1e9f34.png)
Chrome seems to render it the same way. Is this the expected output?

Reply:
@dirkschulze I just realized that the transformation in `::Save` in the Gdiplus port is messed up too. It needs to be concatenated and the way it's happening right now, the matrix is being set directly.